File: TopInterview150/Backtracking/CombinationSum/solution.cpp
class Solution {
    public:
        vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
            vector<vector<int>> result;
            vector<int> comb;

            backtrack(result, comb, candidates, target, 0);
            return result;
        }
        void backtrack(
            vector<vector<int>>& result,
            vector<int>& comb,
            vector<int>& candidates,
            int target,
            int start
        ) {
            if (target == 0)
            {
                result.push_back(comb);
                return;
            }

            for (int i = start; i < candidates.size(); i++)
            {
                if (candidates[i] <= target)
                {
                    comb.push_back(candidates[i]);
                    backtrack(result, comb, candidates, target - candidates[i], i);
                    comb.pop_back();
                }
            }
        }

    };

File: TopInterview150/Backtracking/CombinationSum/solution.py
class Solution(object):
    def backtrack(self, result, comb, candidates, target, start):
        if target == 0:
            result.append(list(comb))
            return
        
        for i in range(start, len(candidates)):
            if candidates[i] <= target:
                comb.append(candidates[i])
                self.backtrack(result, comb, candidates, target-candidates[i], i)
                comb.pop()

    
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        result = []
        comb = []

        self.backtrack(result, comb, candidates, target, 0)
        return result